README.md
# Towards Better Evaluation of GNN Expressiveness with BREC Dataset

## About

This repository is the official implementation of the following paper: [Towards Better Evaluation of GNN Expressiveness with BREC Dataset](https://arxiv.org/abs/2304.07702).

We also provide a Pypi package for simple usage. Please refer to [Pypi package](https://pypi.org/project/brec/).

**BREC**  is a new dataset for GNN expressiveness comparison.
It addresses the limitations of previous datasets, including difficulty, granularity, and scale, by incorporating
400 pairs of various graphs in four categories (Basic, Regular, Extension, CFI).
The graphs are organized pair-wise, where each pair is tested individually to return whether a GNN can distinguish them. We propose a new evaluation method, **RPC** (Reliable Paired Comparisons), with a contrastive training framework.


## Usages

### File Structure

We first introduce the general file structure of BREC:

```bash
├── Data
    └── raw
        └── brec_v3.npy    # unprocessed BREC dataset in graph6 format
├── BRECDataset_v3.py    # BREC dataset construction file
├── test_BREC.py    # Evaluation framework file
└── test_BREC_search.py    # Run test_BREC.py with 10 seeds for the final result
```

To test on BREC, there are four steps to follow:

1. Select a model and go to the corresponding [directory](#directory).
2. [Prepare](#preparation) dataset based on selected model requirements.
3. Check *test_BREC.py* for implementation if you want to test your own GNN.
4. Run *test_BREC_search.py* for final result. Only if no failure in reliability check for all seeds is available.

### Requirements

Tested combination: Python 3.8.13 + [PyTorch 1.13.1](https://pytorch.org/get-started/previous-versions/) + [PyTorch_Geometric 2.2](https://pytorch-geometric.readthedocs.io/en/latest/notes/installation.html)

Other required Python libraries included: numpy, networkx, loguru, etc.

For reproducing other results, please refer to the corresponding requirements for additional libraries.

### <span id="preparation">Data Preparation</span>

Data preparation requires two steps: generate the dataset and arrange it in the correct position.

#### Step 1: Data Generation

We provide zipped data file *BREC_data_all.zip*. You can unzip it for 3 data files in npy format. You can also customize dataset refering to [Customize Dataset](#customize).

For most methods, only *brec_v3.npy* is needed. More detailed requirements on datasets can refer to corresponding implementations.

#### Step 2: Data Arrangement

Replace *$.txt* with *$.npy* in the corresponding *Data/raw* directory.
For most methods, only *brec_v3.txt* is in *Data/raw* directory. Thus replacing *brec_v3.txt* with *brec_v3.npy* is enough.

### <span id="directory">Reproduce Baselines</span>

For baseline results reproduction, please refer to the respective directories:

| Baseline          | Directory                           |
| ----------------- | ----------------------------------- |
| NGNN              | NestedGNN                           |
| DS-GNN            | SUN                                 |
| DSS-GNN           | SUN                                 |
| SUN               | SUN                                 |
| PPGN              | ProvablyPowerfulGraphNetworks_torch |
| GNN-AK            | GNNAsKernel                         |
| DE+NGNN           | NestedGNN                           |
| KP-GNN            | KP-GNN                              |
| KC-SetGNN         | KCSetGNN                            |
| I$^2$-GNN         | I2GNN                               |
| GSN               | GSN                                 |
| Graphormer        | Graphormer                          |
| OSAN              | OSAN                                |
| $\delta$-LGNN(SparseWL) | SparseWL                      |
| SWL               | SWL                                 |
| DropGNN           | DropGNN                             |
| Non-GNN Baselines | Non-GNN                             |
| Your Own GNN      | Base                                |

### Test Your Own GNN

In addition to previous steps in reproducing baselines, implementing *test_BREC.py* is needed.

#### Evaluation Step

To test your GNNs, in addition to previous steps, you need to implement *test_BREC.py* with your model and run (${configs} represents corresponding config usage):

```bash
python test_BREC.py ${configs}
```

*test.py* is the pipeline for evaluation, including four stages:

```bash
1. pre-calculation;

2. dataset construction;

3. model construction;

4. evaluation
```

**Pre-calculation** aims to organize offline operations on graphs.

**Dataset construction** aims to process the dataset with specific operations. BRECDataset is implemented based on InMemoryDataset. It is recommended to use transform and pre_transform to transform the graphs.

**Model construction** aims to construct the GNN.

**Evaluation** implements RPC. With the model and dataset, it will produce the final results.

Suppose your own experiment is done by running *python main.py*. You can easily implement *test_BREC.py* with *main.py*. You can drop the training and testing pipeline in *main.py* and split the rest into corresponding stages in *test.py*.

### <span id="customize">Customize BREC Dataset</span>

Some graphs in BREC may be too difficult for some models, like strongly regular graphs that 3-WL can not distinguish.
You can discard some graphs from BREC to reduce test time.
In addition, the parameter $q$ in RPC can also be adjusted when customizing. Only the *customize* directory is required.

```bash
├── Data     # Original graph file
    └── raw
        ├── basic.npy  # Basic graphs
        ├── regular.npy  # Simple regular graphs
        ├── str.npy   # Strongly regular graphs
        ├── cfi.npy   # CFI graphs
        ├── extension.npy # Extension graphs
        ├── 4vtx.npy  # 4-vertex condition graphs
        └── dr.npy   # Distance regular graphs
├── dataset_v3.py    # Generating brec_v3.npy 
├── dataset_v3_3wl.py # Generating brec_v3_3wl.npy    
└── dataset_v3_no4v_60cfi.py  # Generating brec_v3_no4v_60cfi.npy
```

Using *brec_v3.npy* by running *python dataset_v3.py* is enough for most methods.

For customization, suppose you want to discard distance regular graphs from BREC. You need to delete *dr.npy* related codes in *dataset_v3.py*. The total pair number and the "category-id_range" dictionary should also be adjusted.

"NUM" represent $q$ in RPC, which can be adjusted for a different RPC check.

### Results Demonstration

The 400 pairs of graphs are from four categories: Basic, Regular, Extension, CFI. We further split 4-vertex condition and distance regular graphs from Regular as a separate category. The "category-id_range" dictionary is as follows:

```python
  "Basic": (0, 60),
  "Regular": (60, 160),
  "Extension": (160, 260),
  "CFI": (260, 360),
  "4-Vertex_Condition": (360, 380),
  "Distance_Regular": (380, 400),
```

You can refer to the detailed graph in *customize/Data/raw* for analysis.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).


## [1.0.0] - 2023-08-26

### Added

- Distribute Pypi package at [Pypi](https://pypi.org/project/brec).
- Add CHANGELOG.


## [0.3.0] - 2023-06-14

### Added

- Implement 6 GNN baseines: SparseWL, OSAN, SWL, GSN, DropGNN, Graphormer.
- Implement 2 non-GNN baselines: SPD-WL and GD-WL.
- Add MIT LICENSE.

### Changed

- [Breaking Changes] Update $T^2$-statistics threshold.
- Update guidelines.


## [0.2.0] - 2023-04-22

### Added

- Upload data files in Github.
- Add paper arxiv link at [arxiv](https://arxiv.org/abs/2304.07702)

### Changed

- Update guidelines.

### Fixed

- Correct data order for non-GNNs.


## [0.1.0] - 2023-04-16

### Added

- Release BREC dataset and evaluation framework.
- Implement 10 GNN baselines: NGNN, DE+NGNN, DS-GNN, DSS-GNN, SUN, GNN-AK, KP-GNN, I$^2$-GNN, PPGN and KC-SetGNN.
- Implement 6 non-GNN baselines: 3-WL, $S_3$, $S_4$, $N_1$, $N_2$ and $M_1$. (WL, $S_k$ and $N_k$ also extensible)

brec/evaluator.py
import numpy as np
import torch
import torch_geometric
import torch_geometric.loader
from loguru import logger
import time
from brec.dataset import BRECDataset
from tqdm import tqdm
import os
from torch.nn import CosineEmbeddingLoss
import configparser

NUM_RELABEL = 32
P_NORM = 2
OUTPUT_DIM = 16
EPSILON_MATRIX = 1e-7
EPSILON_CMP = 1e-6
SAMPLE_NUM = 400
EPOCH = 20
MARGIN = 0.0
LEARNING_RATE = 1e-4
THRESHOLD = 72.34
BATCH_SIZE = 16
WEIGHT_DECAY = 1e-4
LOSS_THRESHOLD = 0.2
SEED = 2023


#    learning_rate, weight_decay, batch_size, epoch, loss_threshold):
def evaluate(dataset, model, device, log_path=None, training_config=None):
    """_summary_

    Args:
        dataset (torch_geometric.data.InMemoryDataset): Brec Dataset
        model (torch.nn.Module): Model to be tested
        device (torch.device): Torch device
        logger (loguru.logger): Logger
        log_path (String): log path
        training_config (configparser.ConfigParser): training config

    Returns:
        None: None
    """

    # If you want to test on some simple graphs without permutation outputting the exact same embedding, please use S_epsilon.
    # S_epsilon = torch.diag(
    #     torch.full(size=(OUTPUT_DIM, 1), fill_value=EPSILON_MATRIX).reshape(-1)
    # ).to(device)
    def T2_calculation(dataset, log_flag=False):
        with torch.no_grad():
            loader = torch_geometric.loader.DataLoader(dataset, batch_size=BATCH_SIZE)
            pred_0_list = []
            pred_1_list = []
            for data in loader:
                pred = model(data.to(device)).detach()
                pred_0_list.extend(pred[0::2])
                pred_1_list.extend(pred[1::2])
            X = torch.cat([x.reshape(1, -1) for x in pred_0_list], dim=0).T
            Y = torch.cat([x.reshape(1, -1) for x in pred_1_list], dim=0).T
            if log_flag:
                logger.info(f"X_mean = {torch.mean(X, dim=1)}")
                logger.info(f"Y_mean = {torch.mean(Y, dim=1)}")
            D = X - Y
            D_mean = torch.mean(D, dim=1).reshape(-1, 1)
            S = torch.cov(D)
            inv_S = torch.linalg.pinv(S)
            # If you want to test on some simple graphs without permutation outputting the exact same embedding, please use inv_S with S_epsilon.
            # inv_S = torch.linalg.pinv(S + S_epsilon)
            return torch.mm(torch.mm(D_mean.T, inv_S), D_mean)

    part_dict = {
        "Basic": (0, 60),
        "Regular": (60, 160),
        "Extension": (160, 260),
        "CFI": (260, 360),
        "4-Vertex_Condition": (360, 380),
        "Distance_Regular": (380, 400),
    }
    if log_path is not None:
        logger.remove()
        logger.add(log_path)

    global LEARNING_RATE, WEIGHT_DECAY, LOSS_THRESHOLD, BATCH_SIZE, EPOCH
    if training_config is not None:
        LEARNING_RATE = (
            training_config.getfloat("training", "learning_rate")
            if training_config.has_option("training", "learning_rate")
            else LEARNING_RATE
        )
        WEIGHT_DECAY = (
            training_config.getfloat("training", "weight_decay")
            if training_config.has_option("training", "weight_decay")
            else WEIGHT_DECAY
        )
        LOSS_THRESHOLD = (
            training_config.getfloat("training", "loss_threshold")
            if training_config.has_option("training", "loss_threshold")
            else LOSS_THRESHOLD
        )
        BATCH_SIZE = (
            training_config.getint("training", "batch_size")
            if training_config.has_option("training", "batch_size")
            else BATCH_SIZE
        )
        EPOCH = (
            training_config.getint("training", "epoch")
            if training_config.has_option("training", "epoch")
            else EPOCH
        )

    time_start = time.process_time()

    # Do something
    cnt = 0
    correct_list = []
    fail_in_reliability = 0
    loss_func = CosineEmbeddingLoss(margin=MARGIN)

    for part_name, part_range in part_dict.items():
        logger.info(f"{part_name} part starting ---")

        cnt_part = 0
        fail_in_reliability_part = 0
        start = time.process_time()

        for id in tqdm(range(part_range[0], part_range[1])):
            logger.info(f"ID: {id}")
            model.reset_parameters()
            optimizer = torch.optim.Adam(
                model.parameters(), lr=LEARNING_RATE, weight_decay=WEIGHT_DECAY
            )
            scheduler = torch.optim.lr_scheduler.ReduceLROnPlateau(optimizer)
            dataset_traintest = dataset[
                id * NUM_RELABEL * 2 : (id + 1) * NUM_RELABEL * 2
            ]
            dataset_reliability = dataset[
                (id + SAMPLE_NUM)
                * NUM_RELABEL
                * 2 : (id + SAMPLE_NUM + 1)
                * NUM_RELABEL
                * 2
            ]
            model.train()
            for _ in range(EPOCH):
                traintest_loader = torch_geometric.loader.DataLoader(
                    dataset_traintest, batch_size=BATCH_SIZE
                )
                loss_all = 0
                for data in traintest_loader:
                    optimizer.zero_grad()
                    pred = model(data.to(device))
                    loss = loss_func(
                        pred[0::2],
                        pred[1::2],
                        torch.tensor([-1] * (len(pred) // 2)).to(device),
                    )
                    loss.backward()
                    optimizer.step()
                    loss_all += len(pred) / 2 * loss.item()
                loss_all /= NUM_RELABEL
                # logger.info(f"Loss: {loss_all}")
                if loss_all < LOSS_THRESHOLD:
                    # logger.info("Early Stop Here")
                    break
                scheduler.step(loss_all)

            model.eval()
            T_square_traintest = T2_calculation(dataset_traintest, True)
            T_square_reliability = T2_calculation(dataset_reliability, True)

            isomorphic_flag = False
            reliability_flag = False
            if T_square_traintest > THRESHOLD and not torch.isclose(
                T_square_traintest, T_square_reliability, atol=EPSILON_CMP
            ):
                isomorphic_flag = True
            if T_square_reliability < THRESHOLD:
                reliability_flag = True

            if isomorphic_flag:
                cnt += 1
                cnt_part += 1
                correct_list.append(id)
                logger.info(f"Correct num in current part: {cnt_part}")
            if not reliability_flag:
                fail_in_reliability += 1
                fail_in_reliability_part += 1
            logger.info(f"isomorphic: {isomorphic_flag} {T_square_traintest}")
            logger.info(f"reliability: {reliability_flag} {T_square_reliability}")

        end = time.process_time()
        time_cost_part = round(end - start, 2)

        logger.info(
            f"{part_name} part costs time {time_cost_part}; Correct in {cnt_part} / {part_range[1] - part_range[0]}"
        )
        logger.info(
            f"Fail in reliability: {fail_in_reliability_part} / {part_range[1] - part_range[0]}"
        )

    time_end = time.process_time()
    time_cost = round(time_end - time_start, 2)
    logger.info(f"evaluation time cost: {time_cost}")

    Acc = round(cnt / SAMPLE_NUM, 2)
    logger.info(f"Correct in {cnt} / {SAMPLE_NUM}, Acc = {Acc}")

    logger.info(f"Fail in reliability: {fail_in_reliability} / {SAMPLE_NUM}")
    logger.info(correct_list)
